New Graphics SettingsUI

Armored-Dragon opened the pull request:
Continued from #1240, which is continued from #1125.
Rebased on master!
Replacement Settings UI made to be easily extended and otherwise worked with in the future.

This PR only changes the Graphics UI, leaving all others unchanged.

Armored-Dragon commented:
This is currently missing a setting "3rd Person Camera Clipping". Do not merge this just yet.

ksuprynowicz commented:
I tested it on Windows in VR and Desktop mode. The new design looks very modern and everything works well :)
In my opinion before merging there are two user experience related things there that we should discuss.
I found the arrows next to settings confusing - they seem clickable but they are not. I'm wondering if they are needed at all?:
![obraz](https://github.com/user-attachments/assets/55ef16e9-7d19-4e72-8541-163ce5bb9e21)
Another thing is that since currently in settings there's only "Graphics" option, this may be confusing for users. Would it be possible before merging to add other positions form "Settings" menu there, especially "General" and "Controls"? I mean not redesigning for now, just adding as is.
![obraz](https://github.com/user-attachments/assets/72e86fb7-b9d9-44f6-800c-f5944f111e3d)

Armored-Dragon commented:
The arrows are supposed to work, but they seem to have broken at some point. I will quickly work on adding the missing settings, but just link them in the settings app. This means that this application will act like a link list for all settings.

Armored-Dragon commented:
After some discussion, we decided the best course of action is to include the other setting applications to the listing in the form of links. While this SettingsUI application does not include them itself, these additional categories will allow users to quickly locate the settings section they are looking for.
![image](https://github.com/user-attachments/assets/d7ee0252-9be1-454b-8324-73c0bf4024fb)

JulianGro commented:
Could we wait with merging this until I can give this a proper review next week?
I already see a lot of "doesn't work" comments in the code, and just from the screenshots, it looks like options which have been turned off are greyed out in such a way that it looks like you can't turn them on.

Armored-Dragon commented:
> (I assume we are going to move the new UI from the `scripts` folder to the `interface/resources/qml`, once the new UI is completed.)

My plan right now is to keep all of the components together as much as possible, actual UI (`settings.qml`) with the script. I am wanting to iron out everything in the `qml_widgets` folder so those could be globalized and be used elsewhere. A primary goal for this app is ease of development in the future and I think that having pieces of the app sprawled out everywhere makes development harder. As I understand it, outside of it just being existing practice, having the script UI elements in the recources/qml provides no benefit to this app. Nothing (yet) is meant to be accessed by any other script including the qml widgets.
If I am wrong, please correct me. It might actually have more of a reason considering this is meant to be a primary system app.

On a slightly different but slightly relevant topic, I am toying around with concepts that would allow scripts to declare settings to this app which would allow a centralized place to edit those mentioned settings. This would get some work out of the way for other developers and focus more on the core functionality of their app. This is to say it's possible that by the time this whole Settings rework is done, none of the qml widgets would be globalized or recommended for general use outside of this singular app.

> [ Everything else... ]

Noted, thank you! I should be able to address everything.

JulianGro commented:
> As I understand it, outside of it just being existing practice, having the script UI elements in the recources/qml provides no benefit to this app.

The difference between the resources folder and the scripts folder is that the resources folder gets bundled into a `resources.rcc`, which I am pretty sure increases file system performance (and also just compressed the files). I believe the scripts folder is separate, because the scripting engine has trouble accessing relative paths inside the `resources.rcc`.
But yeah, during development it is indeed easier to have stuff outside the compiled resources, as you don't have to juggle with environment variables to load directly from the source instead of from the compiled resources.

In the future we could improve performance of the compiled resources even more, by pre-compiling the QML and minifying the scripts and JSON files during the resource compilation.

Armored-Dragon commented:
Here is what we are working with now. I think this is a steep improvement. This also gave me a moment to create widgets for the advanced options of settings and add animations to the section when it needs to be displayed or hidden.

![image](https://github.com/user-attachments/assets/27ade91c-7232-4730-82f1-745d1c2ea605)

[Screencast_20250414_014613.webm](https://github.com/user-attachments/assets/4d2f136e-8e6b-4205-985f-17edc24acd39)

HifiExperiments commented:
I think the (somewhat recently added) “Procedural Materials” checkbox may be missing now

JulianGro commented:
A steep improvement indeed.

A little heads up: If your IDE starts running into issues with code completion and/or shows invalid warnings, the [QML_IMPORT_PATH](https://github.com/overte-org/overte/blob/59a6f5aae1fae479d4570226a204bce3c58d099f/CMakeLists.txt#L505) is probably wrong.

More things I noticed:
- The new UI doesn't have any sound effects. I am not sure if our audio files are amazing, but maybe using the old sounds just with a slightly lower volume would be nice? Especially in VR, the sound effects make the UI feel more alive and responsive to me.
- There is no way back from the old settings pages other than closing them. (Maybe we will merge all of the new UI in one go, making this unimportant.)
- The old settings application (which is still used in VR here) has the Graphics settings at a "bad" position.
- There is no scrollbar, giving no indication that you can scroll down. I would suggest either a scrollbar like Create (where the size of the bar changes depending on how much is visible on screen), or a skinny version of that with an Overlay Scrollbar (basically a big handle sticking out of the bar making it easier to grab in VR).
![Screenshotaasa](https://github.com/user-attachments/assets/98601213-9245-41af-b0a8-59f1214d0554)
Picture of the overlay-scrollbar in Ubuntu. You can click the arrows in Ubuntu, but I think just drag controls are better for our use case.
- The General settings open in a separate window in VR. (All the other settings use the tablet just fine.)
- The font seems to abide to system API scaling.

I will add some more pictures from Windows

ksuprynowicz commented:
@Armored-Dragon This looks really good :)

Armored-Dragon commented:
@JulianGro 
### 1: 
Sound effects have been added. [Commit](https://github.com/overte-org/overte/pull/1385/commits/375b507ccbe832b91bd4098ec73675cc94c370cc). 
Hover effects have also been added. [Commit](https://github.com/overte-org/overte/pull/1385/commits/d23ff06b4aca4db885dd15a295ab3a38802905c8).

### 2:
That is just an unfortunate tradeoff with the current design. You mentioned later that it would be best to include everything all in one go. That is the planned fix for this.

### 3:
I am not sure how I can fix this. From my assumption this is the result of a bug in the ordering of the elements of the VR Menu Bar access app on the tablet. I have not had a change to actually look yet. I will be looking into this but it may take a bit to get to the bottom of. 
Context for readers: The position of the VR exclusive app used to access the menu bar has the position of the "Graphics..." option in an incorrect / unintentional place. 

### 4:
Scrollbars have been added.

### 5:
That was intentional. That seems to be the only way to actually open the "General Settings" window without opening additional windows or other weird quirks of the game happening. Opening this window like how other windows are handled resulted in a tablet window inside of another tablet window. A tablet inside of a tablet.
[Docs](https://apidocs.overte.org/Desktop.html#.show)
https://github.com/overte-org/overte/blob/5aa1d5d88348a3bf71e3a07e18932fdb418eb25b/scripts/system/settings/settings.js#L79-L84
The long term fix for this is the same as point 2, replace all setting windows before merging.

### 6:
I attempted a fix by replacing pointSize with pixelSize. Let me know if this worked!


[Screencast_20250419_030726.webm](https://github.com/user-attachments/assets/c5c636cc-2c06-470f-8e8b-b85245e2146f)

(There is no audio in the video but pleeeease trust me the button clicks are there)

JulianGro commented:
> That is just an unfortunate tradeoff with the current design. You mentioned later that it would be best to include everything all in one go. That is the planned fix for this.

Would you like a new-UI branch to merge all the UI stuff into, before merging into master then? Or, since there is very little negatives to far, maybe it would be enough to have the general settings not be at the end of the list of things to do? That way they are ready before the next release (which will probably be a while, since we are doing a lot of huge changes).

Personally, I would love hover effects in the graphics settings and sound effects when hovering. You know, like you did with the list of settings pages. The old UI had hover effects on almost everything.

Is it deliberate that the scroll bar hides itself almost immediately, rather than staying around? Firefox does this as well, and it has pretty much the same issue as having no scrollbar; The issue that you cannot tell if you can scroll down or not.
The old UI had the scroll bar always be visible.

One thing I hadn't noticed before is that the flicking goes quite far:
[scroll.webm](https://github.com/user-attachments/assets/abb789dd-041f-438d-8cb9-df49ed893ec1)
I believe there is a setting for that in QML. The old UI had it completely disabled, but believe we should just limit it a little. It's probably fine in VR, but in desktop mode with a quick scroll-wheel it seems pretty unreasonable. The relevant term for this appears to be boundsBehavior: https://doc.qt.io/qt-5/qml-qtquick-flickable.html#boundsMovement-prop

ksuprynowicz commented:
@Armored-Dragon I fixed the issue with General settings showing in floating window in VR instead of tablet:
https://github.com/Armored-Dragon/overte/pull/148

ksuprynowicz commented:
IMO when General settings issue is fixed, this is already a massive improvement over current settings UI, so should we test, review and merge it and fix the rest in separate PRs?

Armored-Dragon commented:
> @Armored-Dragon I fixed the issue with General settings showing in floating window in VR instead of tablet: [Armored-Dragon#148](https://github.com/Armored-Dragon/overte/pull/148)

Thank you!

Armored-Dragon commented:
> Would you like a new-UI branch to merge all the UI stuff into, before merging into master then?

Right now I don't think that will be a good idea. My idea is that one thing at a time gets merged into master, and if something breaks or is not correct I only have to worry about the one thing. By the end of this whole process everything should be working properly with maybe a few minor nitpicks. I fear if we just merge everything all in one go a lot will break and it would make using Overte a pain for a while, and maybe we'd have to do a few big releases in quick succession just to address these issues that would crop up.

> Personally, I would love hover effects in the graphics settings and sound effects when hovering. You know, like you did with the list of settings pages. The old UI had hover effects on almost everything.

You may be right about the visual hover effects, I can see what I can do about that. The sound effects I don't think will be a good idea. I found it quite annoying having everything make a noise just moving the mouse around a window. 

> Is it deliberate that the scroll bar hides itself almost immediately, rather than staying around? Firefox does this as well, and it has pretty much the same issue as having no scrollbar; The issue that you cannot tell if you can scroll down or not.
The old UI had the scroll bar always be visible.

Yes it was deliberate. I wanted to show that the scrollbar exists and that any page can be scrolled if needed. My idea behind it was I wanted to not block the actual content of the window with the scrollbar but I still wanted it to be obvious that the scrollbar exists. Last night while in bed I thought of a more elegant solution that should solve the scrollbar visibility issue and I will try to implement it when I am next able.

> IMO when General settings issue is fixed, this is already a massive improvement over current settings UI, so should we test, review and merge it and fix the rest in separate PRs?

I would prefer to do more small PRs rather than a single big one here as the different settings windows seem to be embedded at a slightly different depth for each. I think it would be easier to follow, review, and test when what I do is more small scale. My hope is that this PR with just the Graphics settings will be the biggest while the other(s) are just simple rewrites. Depending on the scale, it's possible the other settings pages will together be just a single PR.
Having a "soft launch" of the new Settings UI will also give valuable feedback from the more general audience about when needs more polishing when it comes to the embedded components/widgets. 

I'm perfectly fine with either merging now or waiting until every setting page is done, I will work with either.

ksuprynowicz commented:
> I would prefer to do more small PRs rather than a single big one here as the different settings windows seem to be embedded at a slightly different depth for each.

I think it's indeed the best idea to do it this way. Let me know if it's ready and I'll test and review :)
By Graphics setting issue I just meant one that I sent you PR for with Graphics settings opening in a floating window in VR instead of tablet, it should hopefully be resolved upon merging it.

Armored-Dragon commented:
I had completely forgotten to touch up the numbers element located under "Refresh Rate" "Custom" setting. I changed it to more closely fit the updated design scheme. I may need to come back in at some point and readjust the colors.

ksuprynowicz reviewed (approved):
All looks good :)

ksuprynowicz commented:
I'm testing it currently, and new settings look really nice.
One issue that I noticed is that when you set graphics preset to high, and then disable Rendering Effects toggle, it still shows High instead of Custom. Then after changing to different presets, Rendering Effects stay off.
I think changing Rendering settings toggle should change preset to Custom, and changing preset to Low Power, Medium or High should set Rendering Settings toggle to value in that preset.

Armored-Dragon commented:
I'm going through and making sure that all of the provided settings are up to date, and I just noticed that the "fog" seems to be missing from the current Settings qml dialog. [One of the PRs I found](https://github.com/overte-org/overte/blob/d4ca3d5dc15fda6c50bcd75baeecdb299e02b06e/interface/resources/qml/hifi/dialogs/graphics/GraphicsSettings.qml) has the fog option linked to the "Haze" option. Are fog and haze the same thing?
In either case I can't seem to locate anything relating to "fog" in the whole project.

HifiExperiments commented:
idk who called it fog originally but yeah it’s haze everywhere else in the codebase.  I changed it to haze here (which is also when several of these other toggles were added, definitely want to make sure we don’t lose any): #1053

ksuprynowicz commented:
I just tested it and on the newest version rendering effects setting toggle is now switched automatically on presets, but some of its subitems are not switched to a particular preset sometimes. To reproduce, switch to Custom, disable Local Lights, Ambient Occlusion, Haze and Bloom and then switch to High. Shadows setting will be toggled as intended, but rest of them are not.

Armored-Dragon commented:
> I just tested it and on the newest version rendering effects setting toggle is now switched automatically on presets, but some of its subitems are not switched to a particular preset sometimes. To reproduce, switch to Custom, disable Local Lights, Ambient Occlusion, Haze and Bloom and then switch to High. Shadows setting will be toggled as intended, but rest of them are not.

It looks like Local lights never had a proper toggle, this caused the application to fail at setting that local lights toggle, which made the whole UI update function fail to fully execute.

https://github.com/overte-org/overte/blob/47c7b46b60069deea00a2ad610cd46a8132fa1a4/interface/resources/qml/hifi/dialogs/graphics/GraphicsSettings.qml#L310-L324

I changed the local light toggle to more closely align with what is present in the current GUI.

ksuprynowicz commented:
@Armored-Dragon Thank you! Works perfectly now :)
In my opinion this PR is ready to merge. What are your opinions?

Armored-Dragon commented:
> @Armored-Dragon Thank you! Works perfectly now :) In my opinion this PR is ready to merge. What are your opinions?

I think it is ready. I would hope that any remaining issues that may exist are small and/or inconsequential. It might be best to wait until after this weeks developer meet in case someone has something to bring up.
